### ISSUE TITLE
Provide a callback for when a file is being loaded, a way to provide an always used loader or an Watch option

### ISSUE THREAD

**Issue by OlaviSau** [sass/embedded-host-node#190]
Currently it is very difficult to implement a complete watcher for this project.

This is due to 2 issues I have found so far.

1. Errors in files that are removed from the build tree during the first compilation
2. If a file is changed during compilation - there is no way to detect that.

About 1.
Currently an exception provides the message, sassStack and the span - however the sassStack is in a string format and the imports in there are directly from the files without context.

Example:

```
root.scss
  @import "broken-file.scss"
  
  ...other styles

broken-file.scss
  ...invalid scss
```

This will result in an error and using the span property in the error I could watch the broken-file.scss for changes and compile again once it changes, however if the user would remove the import - I would have no way of knowing since that wasn't in my watch list yet.

There is also no clean way to get loaded urls before a successful compilation.

This only leaves an option to parse the sass stack and manually resolve the files, but that's not ideal as the file resolution code already exists in sass embedded and there is no need to duplicate it.
It also breaks the intent contract - the sass stack is probably not intended for programatic use - thus the format might change to improve readability, but that would break the watcher.

Thus there is a need for one of those 3 options.

If I have a callback - then I can immediately add it to watch and if the file changes before it is loaded - then that one is not an issue.
If I have an always used loader - then I can watch it when I load it.
If there is a native watcher - then I don't need to implement it :)

**Comment by nex3** [comment]
I think the best solution here would probably be to move `loaded_files` from `CompileSuccess` to `CompileResult`, so it's available even when compilation fails. But in the interim, you can also define a custom loader that acts as a filesystem loader while also recording all the files that get touched in the compilation process. See [the spec](https://github.com/sass/sass/blob/main/spec/modules.md#resolving-a-file-url) for details on how filesystem loading should work.

**Comment by nex3** [comment]
Moved to https://github.com/sass/embedded-protocol/issues/100

**Comment by OlaviSau** [comment]
@nex3 I like your idea, but I actually tried the custom loader, but I found that it only uses the loader when it can't find the file using normal means. Thus it doesn't produce the full file list. If that was changed and it would be possible to load files manually in general. 
The reason being is that watch without control over loading is very inefficient as on each compile - all the files are basically loaded every time, even though most changes only involve a single file. I am not sure how much this really affects performance with modern SSD's but it's definitely a possible improvement area.
